services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/NewSemanticUsageValidator.java

```diff
@@ -0,0 +1,253 @@
+    List<String> targetSymbols = targetField.schema().getEnumSymbols();
+    if (targetSymbols.stream().anyMatch(symbol -> !currentSymbols.contains(symbol) && symbol.equals(object))) {
+      return returnTrueAndLogError(
+          String.format("Field %s contains new enum value: %s", formatFieldName(currentField.name()), object));
+    }
+    return false;
+  }
+
+  /**
+   * Check if the value is non-default for the given union field.
+   * If the value is null, it is considered default.
+   * If object is non-null, we need to find the schema for the object inside the union and check the default value.
+   * @param object the value to check
+   * @param currentField the field to check with union type
+   * @return true if the value is non-default, false otherwise
+   */
+  public boolean isNonDefaultValueUnion(Object object, Schema.Field currentField) {
+    if (object == null) {
+      return false;
+    }
+
+    Schema subSchema = findObjectSchemaInsideUnion(object, currentField);
+    if (subSchema == null) {
+      return returnTrueAndLogError(
+          String.format(
+              "Field %s: Cannot find the match schema for value %s from schema union",
+              formatFieldName(currentField.name()),
+              object));
+    }
+
+    Schema.Field newCurrentField = AvroCompatibilityHelper.createSchemaField(currentField.name(), subSchema, "", null);
+    return isNonDefaultValueField(object, newCurrentField);
+  }
+
+  /**
+   * Find the schema for the object inside the union.
+   * @param currentField the field with union type
+   * @param object the object to find the schema
+   * @return the schema for the object if found, null otherwise
+   */
+  public static Schema findObjectSchemaInsideUnion(Object object, Schema.Field currentField) {
+    for (Schema subSchema: currentField.schema().getTypes()) {
+      if (isCorrectSchema(object, subSchema)) {
+        return subSchema;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Check if the value is non-default for the given field.
+   * If the value is null, it is considered default.
+   * Default value of each type are only applicable for non-nested fields.
+   * If the field is nested-field, the default value is always null.
+   *
+   * @param object: the value to check
+   * @param field: the field to check
+   * @return true if the value is non-default, false otherwise
+   */
+  public boolean isNonDefaultValueField(Object object, Schema.Field field) {
+    if (object == null)
+      return false;
+
+    Object fieldDefaultValue = AvroSchemaUtils.getFieldDefault(field);
+    Object typeDefaultValue = getDefaultForType(field.schema().getType());
+
+    if (!Objects.equals(object, typeDefaultValue) && !Objects.equals(object, fieldDefaultValue)) {
+      return returnTrueAndLogError(
+          String.format(
+              "Field: %s contains non-default value. Actual value: %s. Default value: %s or %s",
```

Comment:
Just curious, if default is null, would there be NPE when printing this string?

Reply:
nope. this goes in normally. example http://github.com/linkedin/venice/pull/1574/files#diff-ab9e2fbeeb18564c19f00c706cf6f4659aec783e10fd5447d00ec3718601dec1R289